### DISCUSSION TITLE
For the Python implementation, should qradd and lsqr be re-implemented or are there better Python equivalents?

### DISCUSSION THREAD

**Discussion by nbelakovski** [libprima/prima#35] [Q&A]
I'm working through implementing trstlp and trstlp_sub, and I see that they use a QR factorization which gets updated on every loop and also used to solve a least squares problem. I also noticed the note in linalg.f90 about using native/library implementations for the functions within that module, but I don't see a routine in numpy that would compute least squares using a user-provided Q and R (or R diagonal for that matter) matrices.

The question I'm trying to ask ultimately boils down to this:

Is there a way I can implement trstlp_sub in Python by making use of existing libraries? This comes with the understanding that if we want to use a custom implementation later, we can always do that, but that using existing libraries may speed up the process of getting a first cut at the Python implementation.

**Comment by zaikunzhang** [comment]
Hi @nbelakovski ! I do not think there is an equivalent of `qradd` somewhere. 

As for `lsqr`, there only exist equivalents that do not use user-supplied QR factorization: https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html
